modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java

```diff
@@ -354,43 +429,133 @@ static Result handleDisjunction(List<Query> disjunctions, int minimumShouldMatch
         return new Result(verified, terms);
     }
 
-    static Set<Term> selectTermListWithTheLongestShortestTerm(Set<Term> terms1, Set<Term> terms2) {
+    static Set<QueryTerm> selectTermListWithTheLongestShortestTerm(Set<QueryTerm> terms1, Set<QueryTerm> terms2) {
+        assert terms1 != null || terms2 != null;
         if (terms1 == null) {
             return terms2;
         } else if (terms2 == null) {
             return terms1;
         } else {
-            int terms1ShortestTerm = minTermLength(terms1);
-            int terms2ShortestTerm = minTermLength(terms2);
-            // keep the clause with longest terms, this likely to be rarest.
-            if (terms1ShortestTerm >= terms2ShortestTerm) {
-                return terms1;
+            // Prefer term based extractions over range based extractions:
```

Comment:
Combining both range and term based extraction is tricky. It is like comparing apples with oranges. However it is very likely both appear in the same percolator query. I think therefor at some point we should come up with a way to weight these extractions (based on user configuration) so that all types of extractions can be compared.